sklearn/mixture/_gaussian_mixture_ic.py

```diff
@@ -0,0 +1,776 @@
+            except AssertionError:
+                gm_params["reg_covar"] = _increase_reg(gm_params["reg_covar"])
+                continue
+            # if the code gets here, then the model has been fit with
+            # no errors or singleton clusters
+            if self.criterion == "bic":
+                criter = model.bic(X)
+            else:
+                criter = model.aic(X)
+            break
+
+        results = _CollectResults(model, criter, gm_params, ag_params)
+        return results
+
+    def fit(self, X, y=None):
+        """
+        Fit several Gaussian mixture models to the data.
+        Initialize with agglomerative clustering then
+        estimate model parameters with EM algorithm.
+        Select the best model according to the chosen
+        information criterion.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            List of n_features-dimensional data points. Each row
+            corresponds to a single data point.
+
+        Returns
+        -------
+        self : object
+            Returns an instance of self.
+        """
+
+        self._check_parameters()
+        X = check_array(
+            X, dtype=[np.float64, np.float32], ensure_min_samples=1
+        )
+
+        # check n_components against sample size
+        n_comps = [self.max_components, self.min_components]
+        names = ["max_components", "min_components"]
+        for i in range(len(names)):
+            if n_comps[i] > X.shape[0]:
+                msg = names[i] + "must be <= n_samples, but" + names[i]
+                msg += "= {}, n_samples = {}".format(n_comps[i], X.shape[0])
+                raise ValueError(msg)
+
+        # check if X contains the 0 vector
+        if np.any(~X.any(axis=1)):
+            if "cosine" in self.affinity or self.affinity == "all":
+                if isinstance(self.affinity, np.ndarray):
+                    self.affinity = np.delete(
+                        self.affinity, np.argwhere(self.affinity == "cosine")
+                    )
+                if isinstance(self.affinity, list):
+                    self.affinity.remove("cosine")
+                if self.affinity == "all":
+                    self.affinity = ["euclidean", "manhattan", "none"]
+                warnings.warn(
+                    "X contains a zero vector, will not run cosine affinity."
+                )
+            if self.affinity == "cosine":
+                msg = "X contains a zero vector, cannot run cosine affinity."
+                raise ValueError(msg)
+
+        label_init = self.label_init
+        if label_init is not None:
+            if label_init.size != X.shape[0]:
+                msg = "n_samples must be the same as the length of label_init"
+                raise ValueError(msg)
+
+        param_grid = dict(
+            n_components=range(self.min_components, self.max_components + 1),
+            random_state=[self.random_state],
```

Comment:
@tliu68 [randomness] we might want to consider doing something like [this](https://joblib.readthedocs.io/en/latest/auto_examples/parallel_random_state.html#fixing-the-random-state-to-obtain-deterministic-results) to pass random states down to each GaussianMixture object when using joblib. As is I'm not sure what happens when you pass the same one to each child process.